plugin-cgroupv2/src/k8s/token.rs

```diff
@@ -0,0 +1,140 @@
+use std::{
+    fs::File,
+    io::Read,
+    process::Command,
+    sync::Arc,
+    time::{SystemTime, UNIX_EPOCH},
+};
+
+use anyhow::anyhow;
+use base64::{prelude::BASE64_STANDARD_NO_PAD, Engine};
+use tokio::sync::RwLock;
+
+use super::plugin::TokenRetrieval;
+
+pub struct Token {
+    expiration_time: RwLock<Option<u64>>,
+    token_retrieval: TokenRetrieval,
+    value: RwLock<Option<String>>,
+}
+
+impl Token {
+    pub fn new(token_retrieval: TokenRetrieval) -> Self {
+        Self {
+            expiration_time: RwLock::new(None),
+            token_retrieval,
+            value: RwLock::new(None),
+        }
+    }
+
+    /// Check if the token' lifetime has reached it's expiration or not.
+    async fn is_valid(&self) -> bool {
+        let now_secs = match SystemTime::now().duration_since(UNIX_EPOCH) {
+            Ok(ts) => ts.as_secs(),
+            Err(_) => return false,
+        };
+
+        self.expiration_time.read().await.is_some()
+            && self.value.read().await.is_some()
+            && now_secs < self.expiration_time.read().await.unwrap_or_default() // to prevent any weird panic (won't happen)
```

Comment:
What guarantees that `self.expiration_time` still contains `Some` between line 37 and line 39?

Reply:
God ? 😆

Reply:
I can add a lock